The Wikidata API (and possibly Wikimedia APIs in general) seems to fail when the HTTP Header `X-Requested-With` (or any other `X-...` header, AFAIK) is added. It gives the following error message:

> XMLHttpRequest cannot load [https://www.wikidata.org/w/api.php?origin=*&action=wbgetentities&ids=Q21972834&format=json&languages=en](https://www.wikidata.org/w/api.php?origin=*&action=wbgetentities&ids=Q21972834&format=json&languages=en). Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost' is therefore not allowed access.

As you can see however, the `origin=*` parameter IS present, so I don't think that's the problem. The header has definitely got something to do with it:

<!-- begin snippet: js hide: true console: true babel: false -->

<!-- language: lang-js -->

    console.log('With origin=* param:')
    var url = 'https://www.wikidata.org/w/api.php?origin=*&action=wbgetentities&languages=en&format=json&ids=Q1'

    try {
      var xhr = new XMLHttpRequest()
      xhr.open('GET', url, false)
      xhr.send(null)
      console.log('\tWithout X-Requested-With: succeeded')
    } catch (e) {
      console.log('\tWithout X-Requested-With: failed')
    }

    try {
      var xhr = new XMLHttpRequest()
      xhr.open('GET', url, false)
      xhr.setRequestHeader('X-Requested-With', 'xhr')
      xhr.send(null)
      console.log('\tWith X-Requested-With: succeeded')
    } catch (e) {
      console.log('\tWith X-Requested-With: failed')
    }

    console.log('Without origin=* param:')
    url = 'https://www.wikidata.org/w/api.php?action=wbgetentities&languages=en&format=json&ids=Q1'

    try {
      var xhr = new XMLHttpRequest()
      xhr.open('GET', url, false)
      xhr.send(null)
      console.log('\tWithout X-Requested-With: succeeded')
    } catch (e) {
      console.log('\tWithout X-Requested-With: failed')
    }

    try {
      var xhr = new XMLHttpRequest()
      xhr.open('GET', url, false)
      xhr.setRequestHeader('X-Requested-With', 'xhr')
      xhr.send(null)
      console.log('\tWith X-Requested-With: succeeded')
    } catch (e) {
      console.log('\tWith X-Requested-With: failed')
    }

<!-- end snippet -->

Is this expected behaviour? If it is, is there some solution, apart from simply not adding that header?